students/alexander_boone/lesson05/assignments/mailroom.py

```diff
@@ -0,0 +1,188 @@
+
+    # sort donor data by donation_total
+    def sort_key(donors_data):
+        return donors_data['donation_total']
+
+    donors_sorted = sorted(donors, key=sort_key, reverse=True)
+
+    # print donor data
+    for i in range(len(donors_sorted)):
+        first_name = donors_sorted[i]['first_name']
+        last_name = donors_sorted[i]['last_name']
+        donation_avg = (donors_sorted[i]['donation_total'] /
+                        donors_sorted[i]['donation_count'])
+        print(('{:<25}{:^5}${:>14,.2f}{:^5}{:>10}{:^5}${:>14,.2f}'
+               ).format(first_name + ' ' + last_name, ' ',
+                        donors_sorted[i]['donation_total'], ' ',
+                        donors_sorted[i]['donation_count'], ' ',
+                        donation_avg
+                        )
+              )
+
+    # remain on page until user decides to return to main menu
+    q = 'Z'
+    while q != 'Q' and q != 'q':
+        q = safe_input("\nEnter Q to Quit and return to the main menu: ")
+
+
+def letters_to_all():  # OPTION 3
+    '''Write a letter to every donor and save each one to a
+    file on the disk.'''
+
+    # write a letter to a text file for each donor
+    for donor in donors:
+        with open("{first_name}_{last_name}.txt".format(**donor),
+                  'w+') as new_file:
+            letter_body = ('{first_name} {last_name},\n\nThank you'
+                           + ' for donating ${donation_total}. You '
+                           + 'are so kind.\n\nBest,\n\nLocal Charity'
+                           ).format(**donor)
+            new_file.write(letter_body)
+
+
+def quit_program():  # OPTION 4
+    '''Print exit message and quit the program'''
+    exit_message = "Closing the mailroom for the day..."
+    print(exit_message)
+    sys.exit()
+
+
+def safe_input(string):
+    '''Call input() function safely to handle EOFError
+    and KeyboardInterrupt exceptions.'''
+
+    try:
+        output = input(string)
+    except EOFError:
+        print('EOFError exception encountered.')
+        return None
+    except KeyboardInterrupt:
+        print('KeyboardInterrupt exception encountered.')
+        return None
+    else:
+        return output
+
+
+# main code
+if __name__ == '__main__':
+
+    response_dict = {1: thank_you, 2: create_report, 3: letters_to_all,
+                     4: quit_program}
+    response = 0
+    while response != 4:
+        response = 0
+        # display main menu with options
+        options = ["1. Send a Thank You to a single donor", "2. Create a"
+                   + " Report", "3. Send letters to all donors",
+                   "4. Quit"]
+
+        # get and run user response
+        while int(response) not in [1, 2, 3, 4]:
+            try:
+                print(f"----- Main Menu -----\n{options[0]}\n{options[1]}"
+                      + f"\n{options[2]}\n{options[3]}")
+                try:
+                    response = int(safe_input("Enter a number: "))
```

Comment:
why have safe_input return None, if you can just return `0` and avoid additional logic here?